### PR TITLE
[TIR] Allow converting `BufferRegion` to vectorized `BufferLoad`

### PR DESCRIPTION
**Pull request by cyx-6**

### CODE REVIEW

File: src/ir/expr.cc [hunk 1]
```diff
@@ -49,6 +50,18 @@ PrimExpr PrimExpr::FromObject_(ObjectRef ref) {
...
+    Array<PrimExpr> indices;
+    for (Range r : buffer_region->region) {
+      if (arith::Analyzer().CanProveEqual(r->extent, 1)) {
+        indices.push_back(r->min);
+      } else {
+        indices.push_back(tir::Ramp(r->min, 1, Downcast<IntImm>(r->extent)->value));
+      }
+    }
+    return tir::BufferLoad(buffer_region->buffer, indices);
+  }
```
Comment (junrushao1994) [inline src/ir/expr.cc:64]: This impl needs some small clean-up:

```suggestion
  if (const auto* buffer_region = ref.as<tir::BufferRegionNode>()) {
    Array<PrimExpr> indices;
    indices.reserve(buffer_region->region.size());
    for (const Range& r : buffer_region->region) {
      if (is_one(r->extent)) {
        indices.push_back(r->min);
      } else if (const auto* extent = r->extent.as<IntImmNode>()) {
        indices.push_back(tir::Ramp(r->min, make_const(r->min->dtype, 1), extent->value));
      } else {
        LOG(FATAL) << "ValueError: Cannot convert to BufferLoad: " << ref;
      }
    }
    return tir::BufferLoad(buffer_region->buffer, indices);
  }
```

File: src/tir/ir/expr.cc [hunk 2]
```diff
@@ -828,12 +829,26 @@ TVM_REGISTER_GLOBAL("tir.Call")
...
       Array<PrimExpr> prim_expr_args;
       for (const auto& it : args) {
-        ICHECK(it->IsInstance<runtime::StringObj>() || it->IsInstance<PrimExprNode>())
+        ICHECK(it->IsInstance<runtime::StringObj>() || it->IsInstance<PrimExprNode>() ||
+               it->IsInstance<IterVarNode>() || it->IsInstance<BufferRegionNode>())
             << "Argument " << it << " is not a string or primexpr";
         if (const auto* str = it.as<runtime::StringObj>()) {
           prim_expr_args.push_back(StringImm(str->data));
+        } else if (const auto* expr = it.as<PrimExprNode>()) {
+          prim_expr_args.push_back(GetRef<PrimExpr>(expr));
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:838]: Usually we prefer checking leaf types first

```suggestion
        } else if (const auto* iter_var = it.as<IterVarNode>()) {
          prim_expr_args.push_back(GetRef<IterVar>(iter_var));
```

File: src/tir/ir/expr.cc [hunk 3]
```diff
@@ -828,12 +829,26 @@ TVM_REGISTER_GLOBAL("tir.Call")
...
+        } else if (const auto* br = it.as<BufferRegionNode>()) {
+          BufferRegion buffer_region = GetRef<BufferRegion>(br);
+          Array<PrimExpr> indices;
+          for (Range r : buffer_region->region) {
+            if (arith::Analyzer().CanProveEqual(r->extent, 1)) {
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:843]:
```suggestion
            if (is_one(r->extent, 1)) {
```

File: src/tir/ir/expr.cc [hunk 4]
```diff
@@ -828,12 +829,26 @@ TVM_REGISTER_GLOBAL("tir.Call")
...
+              indices.push_back(r->min);
+            } else {
+              indices.push_back(tir::Ramp(r->min, 1, Downcast<IntImm>(r->extent)->value));
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:846]: ditto, use `make_const` to make sure the dtype is consistent: https://github.com/apache/tvm/pull/12420/files#r945234175

File: src/tir/ir/expr.cc [hunk 5]
```diff
@@ -828,12 +829,26 @@ TVM_REGISTER_GLOBAL("tir.Call")
...
-        ICHECK(it->IsInstance<runtime::StringObj>() || it->IsInstance<PrimExprNode>())
+        ICHECK(it->IsInstance<runtime::StringObj>() || it->IsInstance<PrimExprNode>() ||
+               it->IsInstance<IterVarNode>() || it->IsInstance<BufferRegionNode>())
             << "Argument " << it << " is not a string or primexpr";
         if (const auto* str = it.as<runtime::StringObj>()) {
           prim_expr_args.push_back(StringImm(str->data));
+        } else if (const auto* expr = it.as<PrimExprNode>()) {
+          prim_expr_args.push_back(GetRef<PrimExpr>(expr));
+        } else if (const auto* br = it.as<BufferRegionNode>()) {
+          BufferRegion buffer_region = GetRef<BufferRegion>(br);
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:840]: no need for `GetRef`

```suggestion
        } else if (const auto* buffer_region = it.as<BufferRegionNode>()) {
```

File: src/tir/ir/expr.cc [hunk 6]
```diff
@@ -828,10 +828,24 @@ TVM_REGISTER_GLOBAL("tir.Call")
...
+          prim_expr_args.push_back(GetRef<IterVar>(iter_var)->var);
+        } else if (const auto* br = it.as<BufferRegionNode>()) {
+          Array<PrimExpr> indices;
+          for (Range r : br->region) {
+            if (is_one(r->extent)) {
+              indices.push_back(r->min);
+            } else {
+              indices.push_back(tir::Ramp(r->min, make_const(r->min->dtype, 1),
+                                          Downcast<IntImm>(r->extent)->value));
+            }
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:846]: let's handle `r->extent` in the same way as what we are doing on line 59-63 in `src/ir/expr.cc`

File: src/tir/ir/expr.cc [hunk 7]
```diff
@@ -828,10 +828,25 @@ TVM_REGISTER_GLOBAL("tir.Call")
...
+          prim_expr_args.push_back(GetRef<IterVar>(iter_var)->var);
+        } else if (const auto* br = it.as<BufferRegionNode>()) {
+          Array<PrimExpr> indices;
+          for (Range r : br->region) {
+            if (is_one(r->extent)) {
+              indices.push_back(r->min);
+            } else if (const auto* extent = r->extent.as<IntImmNode>()) {
+              indices.push_back(tir::Ramp(r->min, make_const(r->min->dtype, 1), extent->value));
+            } else {
+              LOG(FATAL) << "ValueError: Cannot convert to BufferLoad: " << br;
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:846]:
```suggestion
              LOG(FATAL) << "ValueError: Cannot convert to BufferLoad: " << GetRef<BufferRegion>(br);
```

**Review by junrushao — APPROVED** [review]
LGTM!

**Pull request merged**